### PR TITLE
Lecture4

### PR DESCRIPTION
**Pull request by artemrakov**
### Solutions for Lecture 4

Спасибо большое за курс <3 

cc @chshersh @vrom911

### CODE REVIEW

File: src/Lecture4.hs [hunk 1]
```diff
@@ -134,8 +143,43 @@ errors. We will simply return an optional result here.
...
+--     helper [] acc = [acc]
+--     helper (x:xs) acc | d == x = acc : helper xs ""
+--                       | otherwise = helper xs (acc ++ [x])
+
+split :: Char -> String -> [String]
+split _ [] = [""]
+split d (x:xs) | x == d  = "" : split d xs
+               | otherwise = (x : y) : ys
+    where
+      (y:ys) = split d xs
```
Comment (artemrakov) [inline src/Lecture4.hs:158]: > Pattern match(es) are non-exhaustive  In a pattern binding: Patterns not matched: []

Получал вот такую ошибку, что-то не знаю. Как это исправить

**Comment by artemrakov** [comment]
Можно еще совету по улучшению скорости

CleanShot 2022-02-14 at 22 24 57@2x

File: src/Lecture4.hs (hunk 1 above)
Comment (chshersh) [reply src/Lecture4.hs:158]: Дело в том, что здесь происходит паттерн-мэтчинг только по одному паттерну, но нет статический гарантий, что функция `split` вернёт не пустой список:

```haskell
(y:ys) = split d xs
```

Поэтому компилятор и выдаёт ворнинг, что эта функция может бросить ошибку в рантайме, если `split d xs`  вернёт пустой список `[]` ⚠️

File: src/Lecture4.hs [hunk 2]
```diff
@@ -134,8 +143,43 @@ errors. We will simply return an optional result here.
...
+
+dropSpaces :: String -> String
+dropSpaces = takeWhile (not . isSpace) . dropWhile isSpace
```
Comment (chshersh) [inline src/Lecture4.hs:161]: You can import this function directly from `Lecture3` module and use it 🙂 
But, you actually don't need to use it. I see that you drop spaces when parsing `TradeType`. You can use `readMaybe` instead, this function already drops spaces 👾

File: src/Lecture4.hs [hunk 3]
```diff
@@ -134,8 +143,43 @@ errors. We will simply return an optional result here.
...
+readTrade :: String -> Maybe TradeType
+readTrade "Buy" = Just Buy
+readTrade "Sell" = Just Sell
+readTrade _ = Nothing
+
+readMaybeNatural :: String -> Maybe Int
+readMaybeNatural str = do
+  n <- readMaybe str
+  guard $ n >= 0
+  pure n
```
Comment (chshersh) [inline src/Lecture4.hs:173]: Nice usage of `do`-syntax! 👏🏻

File: src/Lecture4.hs [hunk 4]
```diff
@@ -134,8 +143,43 @@ errors. We will simply return an optional result here.
...
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow string = case split ',' string of
+                    ["", _, _] -> Nothing
+                    [prod, trade, cost] -> do
+                                           t <- readTrade . dropSpaces $ trade
+                                           c <- readMaybeNatural cost
+                                           pure (Row prod t c)
+                    _ -> Nothing
```
Comment (chshersh) [inline src/Lecture4.hs]: You have an excellent monadic implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow string = do
    [prod, trade, cost] <- Just $ split ',' string
    guard $ prod /= ""
    t <- readMaybe trade
    c <- readMaybeNatural cost
    pure $ Row prod t c
```

File: src/Lecture4.hs [hunk 5]
```diff
@@ -156,7 +200,18 @@ string.
 
 If both strings have the same length, return the first one.
 -}
+
+longestString :: String -> String -> String
+longestString stringA stringB = helper stringA stringB
+  where
+    helper _ [] = stringA
+    helper [] _ = stringB
+    helper (_:xs) (_:ys) = helper xs ys
```
Comment (chshersh) [inline src/Lecture4.hs:208]: That's a nice helper function! 👏🏻 
It's much better and faster than calling `length` function twice on both strings 🚀

File: src/Lecture4.hs [hunk 6]
```diff
@@ -156,7 +200,18 @@ string.
...
+
+-- I know about max but max does work not as we expect in this scenario
```
Comment (chshersh) [inline src/Lecture4.hs:210]: I think `max` won't work here 🤔

File: src/Lecture4.hs [hunk 7]
```diff
@@ -183,9 +238,25 @@ The 'Stats' data type has multiple fields. All these fields have
 instance for the 'Stats' type itself.
 -}
 
-instance Semigroup Stats where
+strictSemiGroupMaybe :: Semigroup a => Maybe a -> Maybe a -> Maybe a
+strictSemiGroupMaybe Nothing a = a
+strictSemiGroupMaybe a Nothing = a
+strictSemiGroupMaybe (Just !a) (Just !b) = Just (a <> b)
```
Comment (chshersh) [inline src/Lecture4.hs]: That's absolutely correct implementation 👍🏻 
You can write it a bit shorter though by pattern matching only on the final result:

```suggestion
strictSemiGroupMaybe a b = case a <> b of
    Nothing -> Nothing
    Just !x -> Just x
```

This implementation also will be lazier in a good way 🦥

File: src/Lecture4.hs [hunk 8]
```diff
@@ -226,7 +316,7 @@ implement the next task.
 -}
 
 combineRows :: NonEmpty Row -> Stats
-combineRows = error "TODO"
+combineRows (x:|xs) = foldl' (\acc row -> acc <> rowToStats row) (rowToStats x) xs
```
Comment (chshersh) [inline src/Lecture4.hs:318]: That's a nice implementation using strict fold 😎

File: src/Lecture4.hs [hunk 9]
```diff
@@ -236,8 +326,30 @@ If there's no some value (for example, there were not "Buy" products),
...
 displayStats :: Stats -> String
-displayStats = error "TODO"
+displayStats stats = unlines content
+  where
+    content =
+      [ "Total positions:       :" ++ " " ++ show (getSum $ statsTotalPositions stats)
+      , "Total final balance    :" ++ " " ++ show (getSum $ statsTotalSum stats)
+      , "Biggest absolute cost  :" ++ " " ++ show (getMax $ statsAbsoluteMax stats)
+      , "Smallest absolute cost :" ++ " " ++ show (getMin $ statsAbsoluteMin stats)
+      , "Max earning            :" ++ " " ++ Data.Maybe.maybe "no value" (show . getMax) (statsSellMax stats)
```
Comment (chshersh) [inline src/Lecture4.hs]: `maybe` comes from `Prelude` so you don't need to use a fully-qualified name here. You can write just `maybe`

```suggestion
      , "Max earning            :" ++ " " ++ maybe "no value" (show . getMax) (statsSellMax stats)
```

**Review by chshersh — APPROVED** [review]
:tada: Fantastic work! Congratulations on finishing all **four** lectures! 🥳 
You're truly a Functional Adept now 🧠 
I hope you had fun :man_dancing: :microphone:

If you would like to receive a pretty certificate about the course completion, please, send your full name and the link to this PR to `haskell.beginners2022@gmail.com` 📧 

If you have time, I would appreciate your feedback on the course:

* [Haskell Beginners 2022 Course Feedback](https://docs.google.com/forms/d/e/1FAIpQLSfCkST64dnDCYkz5mezkVyLfWjeskSxhV0Y_cYxUaJw0U9ZhQ/viewform)

If you like my work, you can [sponsor me on GitHub](https://github.com/sponsors/chshersh) 💖

You can also ask any questions here or anywhere else and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger!

If you want to improve the performance of your solution, the best thing to do would be to get rid of the `String` type and replace it with a more efficient data type. For example, you can use lazy bytestring from the `bytestring` package and read the file lazily. Some functions that will be helpful:

* https://hackage.haskell.org/package/bytestring-0.11.2.0/docs/Data-ByteString-Lazy.html#v:readFile
* https://hackage.haskell.org/package/bytestring-0.11.2.0/docs/Data-ByteString-Lazy-Char8.html#v:lines
* https://hackage.haskell.org/package/bytestring-0.11.2.0/docs/Data-ByteString-Lazy-Char8.html#v:readInt

<hr>

If you would like to continue your journey to Haskell, check out the [Haskell Knowledge Map](https://kowainik.github.io/images/Haskell_Knowledge_Map.png) for the list of topics to explore 🗺️

**Comment by artemrakov** [comment]
Thanks for all the feedback and support :)

File: src/Lecture4.hs (hunk 1 above)
Comment (artemrakov) [reply src/Lecture4.hs:158]: Я думал, что он никогда не может вернуть пустой array т.к  у меня есть `split _ [] = [""]`. Я поэтому туплю что-то

Comment (chshersh) [reply src/Lecture4.hs:158]: Это по реализации так. Но по типу, фунция `split` возвращает обычный список, а у обычного списка может быть два конструктора: пустой и нет.

Если хочется на уровне типов отразить, что функция `split` возвращает непустой список всегда, то можно поменять тип результата (и реализацию соответственно) с `[String]` на `NonEmpty String`.

**Pull request merged**